graviti/utility/attr.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 Graviti. Licensed under MIT License.
+#
+"""Attr related class."""
+
+from itertools import chain
+from typing import (
+    Any,
+    Dict,
+    Generic,
+    Iterable,
+    Iterator,
+    Mapping,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from tensorbay.utility import ReprMixin, ReprType
+
+_T = TypeVar("_T")
+_D = TypeVar("_D")
+
+
+class _AttrDict(Generic[_T], ReprMixin):
+
+    _repr_type = ReprType.MAPPING
+
+    def __init__(self) -> None:
+        self._modules: Dict[str, _AttrDict[_T]] = {}
+        self._data: Dict[str, _T] = {}
+
+    def __len__(self) -> int:
+        return sum(map(len, self._modules)) + len(self._data)
+
+    def __iter__(self) -> Iterator[str]:
+        for key, value in self._modules.items():
+            for suffix in value:
+                yield ".".join((key, suffix))
+
+        yield from self._data
+
+    def __contains__(self, key: object) -> bool:
+        if not isinstance(key, str):
+            return False
+
+        try:
+            self._getitem(key)
+        except KeyError:
+            return False
+
+        return True
+
+    def __getitem__(self, key: str) -> _T:
+        try:
+            return self._getitem(key)
+        except KeyError:
+            return self.__missing__(key)
+
+    def __setitem__(self, key: str, value: _T) -> None:
+        try:
+            prefix, suffix = key.split(".", 1)
+        except ValueError:
+            if key in self._modules:
+                raise KeyError("Key prefix conflict") from None
```

Comment:
Setting an entire module is not allowed, but setting an item in one module is accepted?

Reply:
`AttrDict` does not provide the ability to set non-leaf node manually,
`__setitem__` will treat all values input from user as leaf node even it is another `AttrDict`